### PR TITLE
Add local cache for transformPatternToRegex

### PR DESCRIPTION
**Pull request by jdeniau**
This little cache helps us save 1,5 second on one file containing 2200 lines and "saved" nearly one billion call to `preg_match`

### CODE REVIEW

**Comment by jdeniau** [comment]
@ciaranmcnulty changed

**Comment by ciaranmcnulty** [comment]
Looks like a cool optimisation - my only worry would be if there's an adverse affect on memory usage for suites which have a large number of _unique_ step defs - any ideas how to profile it?

**Comment by jdeniau** [comment]
In the test I mention, I profiled with blackfire, I could only watch `supportsDefinitionAndArgument` wich call `transformPatternToRegexp` (because in the optimized version, the call disapear, I think blackfire does not display "small" calls).

Well, the memory usage for `supportsDefinitionAndArgument` "bumped" from 45k to 76k. For the record the global memory usage for this test is 1080M in both case (0.004% → 0.007%), so I think it's pretty anectdotic.

As a matter of fact, I think we do more reuse regexp more than we create new ones, especially with tools like `behatch/context`

File: src/Behat/Behat/Definition/Pattern/PatternTransformer.php [hunk 1]
```diff
@@ -68,11 +71,16 @@ public function generatePattern($type, $stepText)
...
-                return $policy->transformPatternToRegex($pattern);
+        if (!isset($this->patternToRegexpCache[$pattern])) {
+            foreach ($this->policies as $policy) {
+                if ($policy->supportsPattern($pattern)) {
+                    $this->patternToRegexpCache[$pattern] = $policy->transformPatternToRegex($pattern);
+                    break;
+                }
             }
         }
+        return $this->patternToRegexpCache[$pattern];
```
Comment (ciaranmcnulty) [inline src/Behat/Behat/Definition/Pattern/PatternTransformer.php]: you should only return it if it's set (the loop maybe got to the end and didn't find anything)

Comment (ciaranmcnulty) [reply src/Behat/Behat/Definition/Pattern/PatternTransformer.php]: Note the unreachable exception after this - you could move the exception up and put a condition around it

Comment (ciaranmcnulty) [reply src/Behat/Behat/Definition/Pattern/PatternTransformer.php]: The fact the tests didn't catch this means there's one missing :-)

Comment (jdeniau) [reply src/Behat/Behat/Definition/Pattern/PatternTransformer.php]: 🤦‍ Of course, it's fixed, and I added a test about it

**Comment by jdeniau** [comment]
@ciaranmcnulty I do not really know how to fix the scrutinizer test without increasing the overall complexity.

I could add some private functions but it seems kind of really complex for that case, WDYT ?

**Comment by jakzal** [comment]
This would normally be solved by decoration, but we don't have an interface here ;)

I'd indeed extract some private methods to make the public method more concise and readable, i.e.:

```php
     public function transformPatternToRegex($pattern)
    {
        if (!isset($this->patternToRegexpCache[$pattern])) {
            $this->patternToRegexpCache[$pattern] = $this->transformPatternToRegexWithSupportedPolicy($pattern);
        }

        return $this->patternToRegexpCache[$pattern];
    }

    private function transformPatternToRegexWithSupportedPolicy(string $pattern): string
    {
        foreach ($this->policies as $policy) {
            if ($policy->supportsPattern($pattern)) {
                return $policy->transformPatternToRegex($pattern);
            }
        }

        throw new UnknownPatternException(sprintf('Can not find policy for a pattern `%s`.', $pattern), $pattern);
    }
```

**Comment by jakzal** [comment]
Sorry... I forgot that Behat still supports php 5, and my example used php 7 constructs. Also, I haven't tested the "rafactoring". Did it in my head ;)

File: tests/Behat/Tests/Definition/Pattern/PatternTransformerTest.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+use Behat\Behat\Definition\Pattern\PatternTransformer;
+use Behat\Behat\Definition\Pattern\Policy\PatternPolicy;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * Class PatternTransformerTest
+ * @author Julien Deniau <julien.deniau@mapado.com>
+ */
+class PatternTransformerTest extends TestCase
```
Comment (everzet) [inline tests/Behat/Tests/Definition/Pattern/PatternTransformerTest.php]: Bonus points for having integration test!

File: src/Behat/Behat/Definition/Pattern/PatternTransformer.php [hunk 3]
```diff
@@ -26,6 +26,8 @@
      */
     private $policies = array();
 
+    private $patternToRegexpCache = array();
```
Comment (everzet) [inline src/Behat/Behat/Definition/Pattern/PatternTransformer.php]: Ideally needs a typehint.

**Comment by ciaranmcnulty** [comment]
Thanks, @jdeniau!

**Pull request merged**